### PR TITLE
Implementation of FluxLimitRate

### PR DESCRIPTION
**Pull request by OlegDokuka**
This PR provides an implementation of Queue-free Rate Limiter (see #1841)

### CODE REVIEW

**Comment by simonbasle** [comment]
Any planned progress on this one @OlegDokuka ?

**Comment by OlegDokuka** [comment]
@simonbasle I need to write up the docs. I have been waiting for your initial prereview to finalize it. If it looks good, going to push immediately the rest of missing docs! 🤓👨‍🔧

File: reactor-core/src/main/java/reactor/core/publisher/FluxLimitRate.java [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+                                PRODUCED.addAndGet(this, -produced);
+                            }
+
+                            missed = WIP.addAndGet(this, -missed);
+                            if (missed == 0) {
+                                break;
+                            }
+                        }
+
+                        produced = this.produced;
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxLimitRate.java]: would it make sense to always make these two variable refresh at the beginning of the loop? (at least right after the first `if (unboundedRequest || unboundedHighTide)`)

File: reactor-core/src/main/java/reactor/core/publisher/FluxLimitRate.java [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+                        r = this.requested;
+                    }
+                    else {
+                        missed = w;
+                        r = this.requested;
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxLimitRate.java]: not needed anymore if we do the variable refresh at the beginning of the loop

File: reactor-core/src/main/java/reactor/core/publisher/FluxLimitRate.java [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+                    }
+                    else if (pending != highTide) {
+                        deliver = Math.min(highTide - pending, r);
+                    }
+
+                    int w = wip;
+                    if (missed == w) {
+                        if (deliver > 0) {
+                            if (produced != 0) {
+                                this.pending = highTide - produced + (int) deliver;
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxLimitRate.java]: The many variables makes the logic hard to follow here.

Let's say we've got `produced == 12, r == 100`. That would mean `deliver == min(12, 100) == 12 == produced`. So here `pending = highTide - 12 + 12`.

Similarly below, if we reach line 132 it means `deliver > 0 && produced == 0`, so `deliver == highTide - pending` and thus line 132 boils down to `this.pending = pending + hightTide - pending = highTide`. So these computations only really make sense when `deliver == r` (ie r is smaller than the produced/hightide-pending).

Maybe that case of `deliver == r` should stand out a bit more? Something like:

```java
if (deliver > 0) {
  if (deliver == r) {
    int smallRequest = deliver; //for readability only, could be replaced with `deliver` + a comment
    if (produced != 0) {
      this.pending = highTide - produced + smallRequest;
      PRODUCED.addAndGet(this, -produced);
    }
    else {
      this.pending += smallRequest;
    }
  }
  else {
    this.pending = highTide;	
    if (produced != 0) {
      PRODUCED.addAndGet(this, -produced);
    }
  }
  Operators.addCap(REQUESTED, this, -deliver);

  missed = WIP.addAndGet(this, -missed);
  if (missed == 0) {
    s.request(deliver);
    break;
  }
}
```

File: reactor-core/src/main/java/reactor/core/publisher/FluxLimitRate.java [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+            if (WIP.getAndIncrement(this) != 0) {
+                return;
+            }
+
+
+            final Subscription s = this.s;
+            final int highTide = this.highTide;
+
+            int missed = 1;
+            int pending = this.pending;
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxLimitRate.java]: I think the `pending`, `produced` and `deliver` variable could benefit from either a renaming or a comment about what they represent, since several operations are performed on them that are not trivial to follow.

**Comment by OlegDokuka** [comment]
Alright, I should get back to this PR shortly

**Comment by maxim-bandurko-lsvt** [comment]
@OlegDokuka @simonbasle Proposing to implement additional variation of limitRate based on Oleg's enhancement. Ability to pass BiFunction that returns Publisher and exposes 2 variables that can be passed to this publisher originator: long offset and long limit. Limit Rate will be passing data from Publisher in batches.

This enhancement will be beneficial, as allows to control how many items were produced, and how many items can be produced more by passed stream reference that BiFunction returns.

This will allow to have different use cases, but most commonly used will be streams that deals with databases, as having limit and offset values calculations that come from Flux signals, will allow to use them inside database queries and split process of data retrieval to rate limited batches.

Also that can be the first step for Spring Data in implementation of Pageable, that listens to native Flux signals. In case of WebFlux, that mostly uses Pageable from PageRequest, it may allow to change logic to native Flux requests qty. But in case of RSocket, with persistent streams, this new Pageable that right now can be generated from limitRate, will allow be utilized right before streams that deals with databases. That will secure server from breaking because of non-controllable high amount originated request numbers that may overload database usage.

As I mentioned in other issues, I already have such my custom operator built based on limitRate, and it completely simplifies injection of Pageable generation for databases access. And it will be nice to have it natively inside Reactor.

Thank you.

**Comment by simonbasle** [comment]
@maxim-bandurko-lsvt can you open an issue (`enhancement` type) to discuss the design and implementation of such an operator / variant of `limitRate` ? pursuing a generic solution the pagination problem sounds interesting. please also ping @mp911de as I'm sure he'll have some insight

**Comment by maxim-bandurko-lsvt** [comment]
@simonbasle Sure. Will open it. Thanks!

**Comment by chemicL** [comment]
The last interaction here was almost 4 years ago. Closing due to inactivity. Please create a new PR targeting the latest codebase in case you feel this is still applicable.
